Add Context menu option to copy link to currently selected text.

megangardner opened the pull request:
#### a204337cb0ff2df80b628fb8415ee6a39ceb4b2f
<pre>
Add Context menu option to copy link to currently selected text.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269400">https://bugs.webkit.org/show_bug.cgi?id=269400</a>
<a href="https://rdar.apple.com/113158483">rdar://113158483</a>

Reviewed by Ryosuke Niwa.

As discussed in  <a href="https://wicg.github.io/scroll-to-text-fragment/#generating-text-fragment-directives">https://wicg.github.io/scroll-to-text-fragment/#generating-text-fragment-directives</a>
Which is non-normative, currently, we can generated fragment directives by
looking at the text that is highlighted and the surrounding text.
This is a first pass implementation and there are several edge cases that still need to be
worked out, but this is a place to start.

Currently, for short pieces of text, we look back up to 3 words, in the same node, in each direction for
context. And if we have a large piece of text, we take the first three words at the beginning and end of
the content. Medium sized pieces of text are just denoted completely.

This does not give enough for situations where there is lots of repeated text, or text that is
repeated several times across a page, but it covers most cases.

This will be expanded with tests and additional work on know edge cases.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeFragmentDirectiveSpecialCharacters):
* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/URLParser.cpp:
(WTF::isSpecialCharacterForFragmentDirective):
(WTF::URLParser::isSpecialCharacterForFragmentDirective):
* Source/WTF/wtf/URLParser.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp: Added.
(WebCore::FragmentDirectiveGenerator::FragmentDirectiveGenerator):
(WebCore::previousWordsFromPosition):
* Source/WebCore/dom/FragmentDirectiveGenerator.h: Added.
(WebCore::FragmentDirectiveGenerator::urlWithFragment const):
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::shouldEnableCopyLinkToHighlight const):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/page/ContextMenuController.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::urlForSelectedText):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/cocoa/LocalizedStringsCocoa.mm:
(WebCore::contextMenuItemTagCopyLinkToHighlight):
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::copyLinkToHighlight):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldAllowHighlightLinkCreation]):
(-[WKContentView buildMenuForWebViewWithBuilder:]):
(-[WKContentView appHighlightMenu]):
(-[WKContentView scrollToTextFragmentGenerationMenu]):
(-[WKContentView shouldAllowAppHighlightCreation]): Deleted.
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::copyLinkToHighlight):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/274919@main">https://commits.webkit.org/274919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58fe0e1b3a82d6cb916292a93e663fb6c995cc30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40377 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42755 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42923 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36466 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42684 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33490 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40951 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16337 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34811 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14074 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14131 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44200 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33830 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36599 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39841 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40003 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12439 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38119 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16811 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47013 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9060 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16860 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9678 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16455 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -329,6 +329,7 @@ template<typename CharacterType> ALWAYS_INLINE static bool isInUserInfoEncodeSet
 template<typename CharacterType> ALWAYS_INLINE static bool isPercentOrNonASCII(CharacterType character) { return !isASCII(character) || character == '%'; }
 template<typename CharacterType> ALWAYS_INLINE static bool isSlashQuestionOrHash(CharacterType character) { return character <= '\\' && characterClassTable[character] & SlashQuestionOrHash; }
 template<typename CharacterType> ALWAYS_INLINE static bool isValidSchemeCharacter(CharacterType character) { return character <= 'z' && characterClassTable[character] & ValidScheme; }
+template<typename CharacterType> ALWAYS_INLINE static bool isSpecialCharacterForFragmentDirective(CharacterType character) { return character == 0x2C || character == 0x2D; }
```
maybe add a comment clarifying what characters these actually are?

hortont424 replied:
or use character literals? ',' and '-' are much more clear

megangardner replied:
Yeah, that's a clear improvement.

rr-codes commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2024

rr-codes commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+static String previousWordsFromPosition(uint numberOfWords, VisiblePosition& startPosition)
+{
+    auto previousPosition = startPosition;
+    while (numberOfWords--) {
```
isn’t this an infinite loop if numberOfWords is unsigned?

megangardner replied:
No, it hits zero and exits out, but it's probably not the best structure.

rniwa commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+
+    if (textFromRange.length() >= maxCharacters) {
+        String startText = percentEncodeFragmentDirectiveSpecialCharacters(nextWordsFromPosition(3, visibleStartPosition));
+        String endText = percentEncodeFragmentDirectiveSpecialCharacters(previousWordsFromPosition(3, visibleEndPosition));
```
We need to think about what happens with Chinese / Japanese where by default every character as a distinct word.

rniwa commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -1425,6 +1435,19 @@ void ContextMenuController::addDebuggingItems()
 
+bool ContextMenuController::shouldEnableCopyLinkToHighlight() const
+{
+    auto* frame = m_context.hitTestResult().innerNonSharedNode()->document().frame();
```
Please use RefPtr.

aprotyas reviewed (commented):
Very cool. LGTM

smfr commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+{
+    generateFragmentDirective(textFragmentRange);
+
+    m_isValid = true;
```
Set this via the initializer in the .h file

smfr commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+    m_isValid = true;
+}
+
+static String previousWordsFromPosition(uint numberOfWords, VisiblePosition& startPosition)
```
We never use `uint`. Just `unsigned` ?

alanbaradlay replied:
or preferably `size_t`

smfr commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+    return range->toString().trim(isHTMLSpaceButNotLineBreak);
+}
+
+static String nextWordsFromPosition(uint numberOfWords, VisiblePosition& startPosition)
```
`uint` -> `unsigned`

smfr commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -0,0 +1,147 @@
+// https://wicg.github.io/scroll-to-text-fragment/#generating-text-fragment-directives
+void FragmentDirectiveGenerator::generateFragmentDirective(const SimpleRange& textFragmentRange)
+{
+    LOG_WITH_STREAM(TextFragment, stream << " generateFragmentDirective: ");
```
I feel like you planned to log the url at the end.

smfr commented on Source/WebCore/dom/FragmentDirectiveGenerator.h:
```diff
@@ -0,0 +1,57 @@
+public:
+    WEBCORE_EXPORT explicit FragmentDirectiveGenerator(const SimpleRange&);
+
+    const URL urlWithFragment() const { return m_urlWithFragment; };
```
No `const` at the start; you're returning a URL by value.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4224,6 +4225,15 @@ bool Page::shouldDisableCorsForRequestTo(const URL& url) const
     });
 }
 
+URL Page::urlForSelectedText()
```
Can this function be `const`? I think this function should also be called `fragmentDirectiveURLForSelectedText()`

smfr commented on Source/WebCore/platform/cocoa/LocalizedStringsCocoa.mm:
```diff
@@ -52,6 +52,11 @@ String contextMenuItemTagAddHighlightToNewQuickNote()
 
+String contextMenuItemTagCopyLinkToHighlight()
+{
+    return WEB_UI_STRING("Copy Link To Highlight", "Copy link to highlight context menu item");
```
No sure if the "To" should be capitalized.

pxlcoder replied:
I don't think it should be capitalized. Similarly, "Add to Quick Note" is not.

megangardner committed:
Add Context menu option to copy link to currently selected text.
https://bugs.webkit.org/show_bug.cgi?id=269400
rdar://113158483

Reviewed by Ryosuke Niwa.

As discussed in  https://wicg.github.io/scroll-to-text-fragment/#generating-text-fragment-directives
Which is non-normative, currently, we can generated fragment directives by
looking at the text that is highlighted and the surrounding text.
This is a first pass implementation and there are several edge cases that still need to be
worked out, but this is a place to start.

Currently, for short pieces of text, we look back up to 3 words, in the same node, in each direction for
context. And if we have a large piece of text, we take the first three words at the beginning and end of
the content. Medium sized pieces of text are just denoted completely.

This does not give enough for situations where there is lots of repeated text, or text that is
repeated several times across a page, but it covers most cases.

This will be expanded with tests and additional work on know edge cases.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeFragmentDirectiveSpecialCharacters):
* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/URLParser.cpp:
(WTF::isSpecialCharacterForFragmentDirective):
(WTF::URLParser::isSpecialCharacterForFragmentDirective):
* Source/WTF/wtf/URLParser.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp: Added.
(WebCore::FragmentDirectiveGenerator::FragmentDirectiveGenerator):
(WebCore::previousWordsFromPosition):
* Source/WebCore/dom/FragmentDirectiveGenerator.h: Added.
(WebCore::FragmentDirectiveGenerator::urlWithFragment const):
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::populate):
(WebCore::ContextMenuController::shouldEnableCopyLinkToHighlight const):
(WebCore::ContextMenuController::checkOrEnableIfNeeded const):
* Source/WebCore/page/ContextMenuController.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::urlForSelectedText):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/ContextMenuItem.cpp:
(WebCore::isValidContextMenuAction):
* Source/WebCore/platform/ContextMenuItem.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/platform/cocoa/LocalizedStringsCocoa.mm:
(WebCore::contextMenuItemTagCopyLinkToHighlight):
* Source/WebKit/Shared/API/c/WKContextMenuItemTypes.h:
* Source/WebKit/Shared/API/c/WKSharedAPICast.h:
(WebKit::toAPI):
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiers.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKMenuItemIdentifiersPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::copyLinkToHighlight):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView shouldAllowHighlightLinkCreation]):
(-[WKContentView buildMenuForWebViewWithBuilder:]):
(-[WKContentView appHighlightMenu]):
(-[WKContentView scrollToTextFragmentGenerationMenu]):
(-[WKContentView shouldAllowAppHighlightCreation]): Deleted.
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::menuItemIdentifier):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::copyLinkToHighlight):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/274919@main
